src/util/threadpool.rs

```diff
@@ -0,0 +1,618 @@
+        // txn21, txn22 }. During [2*sleep_duration,3*sleep_duration],the running
+        // tasks should be {txn31, txn32}
+        assert_eq!(rx.recv_timeout(recv_timeout_duration).unwrap(), group1);
+        assert_eq!(rx.recv_timeout(recv_timeout_duration).unwrap(), group1);
+        let mut group2_num = 0;
+        let mut group1_num = 0;
+        for _ in 0..4 {
+            let group = rx.recv_timeout(recv_timeout_duration).unwrap();
+            if group == group1 {
+                group1_num += 1;
+                continue;
+            }
+            assert_eq!(group, group2);
+            group2_num += 1;
+        }
+        assert_eq!(group1_num, 2);
+        assert_eq!(group2_num, 2);
+        assert_eq!(rx.recv_timeout(recv_timeout_duration).unwrap(), group3);
+        assert_eq!(rx.recv_timeout(recv_timeout_duration).unwrap(), group3);
+    }
+
+    #[test]
+    fn test_get_task_count() {
+        let name = thd_name!("test_get_task_count");
+        let concurrency = 1;
+        let mut task_pool = ThreadPool::new(name, concurrency, BigGroupThrottledQueue::new(1));
+        let (tx, rx) = channel();
+        let sleep_duration = Duration::from_millis(50);
+        let recv_timeout_duration = Duration::from_secs(2);
+        let group_num = 4;
+        let mut task_num = 0;
+        for gid in 0..group_num {
+            let txer = tx.clone();
+            task_pool.execute(gid, move || {
+                sleep(sleep_duration);
+                txer.send(gid).unwrap();
+            });
+            task_num += 1;
+            assert_eq!(task_pool.get_task_count(), task_num);
+        }
+
+        for gid in 0..group_num {
+            assert_eq!(gid, rx.recv_timeout(recv_timeout_duration).unwrap());
+            // To make sure the task was finished and count has been changed.
+            sleep(sleep_duration / 2);
+            task_num -= 1;
+            assert_eq!(task_pool.get_task_count(), task_num);
+        }
+        task_pool.stop().unwrap();
+    }
+
+    #[test]
+    fn test_fair_group_queue() {
+        let max_pending_task_each_group = 2;
+        let mut queue = BigGroupThrottledQueue::new(max_pending_task_each_group);
+        // Push 4 tasks of `group1` into queue
+        let group1 = 1001;
+        let mut id = 0;
+        for _ in 0..4 {
+            let task = Task::new(id, group1, move || {});
+            id += 1;
+            queue.push(task);
+        }
+
+        // Push 2 tasks of `group2` into queue.
+        let group2 = 1002;
+        for _ in 0..2 {
+            let task = Task::new(id, group2, move || {});
+            id += 1;
+            queue.push(task);
+        }
+        // Push 2 tasks of `group3` into queue.
+        let group3 = 1003;
+        for _ in 0..2 {
+            let task = Task::new(id, group3, move || {});
+            id += 1;
+            queue.push(task);
+        }
+        // queue:g1, g1, g1, g1, g2, g2, g3, g3
```

Comment:
These operations are too long to read. It's better to separate them into blocks. You could write the code to cover these rules in the PR description, one by one:

> it schedules according to the following rules:
>
> 1. Find out which group has a running number that is smaller than that of group_concurrency_on_busy.
> 2. If more than one group meets the first point, run the one who comes first.
>
> If no group meets the first point, choose according to the following rules:
> 
> 1. Select the groups with the least running tasks.
> 2. If more than one group meets the first point, choose the one whose task comes first (with the minimum task's ID).

The "If no group meets the first point" part is never tested in this test case.